### ISSUE TITLE
17. BERT(Bidirectional Encoder Representations from Transformers)

### ISSUE THREAD

**Issue by p-jiho** [p-jiho/natural-language#17]
- BERT : 트렌스포머 계열의 사전 훈련된 언어모델 중 하나

#### 1) 사전 훈련된 워드 임베딩
- 임베딩 사용 방법
    - 임베딩 층을 랜덤 초기화하여 처음부터 학습하는 방법
    - 사전에 학습된 임베딩 벡터를 가져와 사용하는 방법
- 두 방법의 단점 : 문맥을 고려하지 못해 다의어나 동음이의어를 구분하지 못한다.
- 사전 훈련된 언어 모델을 사용하므로써 극복 가능 -> 문맥을 파악할 수 있음

#### 2) 사전 훈련된 언어 모델
- 우선 LSTM 언어 모델을 학습한다
- 언어 모델 : 이전 단어로 다음 단어를 예측하므로 별도의 레이블이 필요하지 않음
- 사전 훈련 워드 임베딩, 사전 훈련 언어 모델의 강점 : 학습 전 별도 레이블 지정이 필요없다.
- ELMo : 순방향, 역방향 언어모델을 각각 학습한 후 임베딩 값을 얻는다.
- 트렌스포머의 디코더 : LSTM 언어 모델 처럼 순차적으로 다음 단어 예측
- Open AI : 트랜스포머 디코더로 12개의 층을 쌓아 학습시킨 언어 모델 GPT-1을 만들었다.
- 양방향 언어모델 : 순방향으로 예측하는 시간에 역방향 언어 모델도 정보를 받고 있어 예측할 단어를 미리 관측하게 될 수 있다.  -> 언어모델은 양방향으로 구현하지 않는다.
- 대안 : ELMo는 두 개의 단방향 언어 모델을 준비해 학습하였다.

#### 3) 마스크드 언어 모델
- 입력 텍스트의 단어 집합의 15%의 단어를 랜덤으로 마스킹
- 마스킹 된 단어를 예측한다. -> 문장 중간에 난 구멍

<br></br>

## 2. 버트(BERT)
- 2018년 구글에서 공개한 사전 훈련 모델
- `pip install transformers`

#### 1) BERT의 개요
- 트랜스포머를 이용해 구현. 레이블이 없는 텍스트 데이터로 훈련.
- 파인 튜닝 : 파라미터 재조정을 위해 기존 모델을 추가 훈련하는 과정

#### 2) BERT의 크기
- BERT의 구조 : 트랜스포머의 인코더를 쌓아올린 구조
- Base 버전보다 Large 버전이 인코더 층 수, d_model의 크기, 셀프 어텐션 헤드의 수가 더 크다.
- Large 버전이 최대 성능을 보여줄 수 있다.

#### 3) BERT의 문맥을 반영한 임베딩
- ELMo나 GPT-1과 마찬가지로 문맥을 반영한 임베딩을 사용
- BERT의 입력 : 임베딩 층을 지난 임베딩 벡터. d_model은 768로 지정되었으므로 768 차원의 임베딩 벡터.
- BERT의 연산을 거친 후 출력 임베딩 : 문맥을 모두 참고한 임베딩
- 12개의 층을 모두 지난 후 출력 임베딩을 얻는다. 각 층에서 모든 단어를 참고하는 연산이 각각 이루어진다.
- BERT의 각 층은 멀티 헤드 셀프 어텐션과 포지션 와이즈 피드 포워드 신경망을 수행한다. 즉, 셀프 어텐션이다.

#### 4) BERT의 서브워드 토크나이저 : WordPiece
- 서브워드 토크나이저를 사용한다. : WordPiece
    -> BPE와 조금 다름. 하지만 서브워드를 병합해가는 방식은 동일
- 자주 등장하는 단어 -> 단어 집합에 추가, 자주 등장하지 않는 단어 -> 서브워드로 분리
- BERT의 토큰화
   - 토큰이 단어 집합에 존재 -> 분리하지 않음
   - 토큰이 단어 집합에 존재X -> 분리 후 첫번째를 제외한 서브워드 앞에 "##"을 붙임
- BERT 토크나이저 생성 
   - `from transformers import BertTokenizer`
   - `BertTokenizer.from_pretrained("bert-base-uncased")`
- 토큰화 : `tokenizer.tokenize`
- BERT 내의 특정 단어 조회 : `tokenizer.vocab['단어']`

#### 5) 포지션 임베딩
- 포지션 임베딩 : 사인, 코사인 함수로 얻는 포지서녈 인코딩과 달리 학습을 통해 얻는다.
- 각 단어의 임베딩 벡터에 문장의 길이가 n이면 n개의 포지션 임베딩 벡터를 생성해 각각 더해준다.
- BERT는 문장 최대 길이가 512이므로 512개의 포지션 임베딩 벡터 존재
- 단어 벡터를 위한 임베딩 층, 포지션 벡터를 위한 임베딩 층, 총 두개가 존재한다.

#### 6) BERT의 사전 훈련
- ELMo : 순방향과 역방향 LSTM을 각각 훈련 시킨다.
- GPT-1 : 트렌스포머의 디코더를 단방향 언어모델로 예측하는 방식
- BERT : 마스크드 언어 모델을 통해 양방향성을 얻음
     - 1. 마스크드 언어 모델    2. 다음 문장 예측(NSP)

1. 마스크드 언어 모델(MLM)
- 입력 텍스트의 15% 단어를 랜덤 마스킹 -> 마스킹된 단어들을 예측
- 15%의 단어에 대한 규칙
     - 80%는 [MASK]로 변경
     - 10%는 랜덤으로 단어 변경
     - 10%는 동일하게 둠
     -> [MASK]만 사용할 경우 이 토큰이 파인 튜닝 단계에서 나타나지 않아 사전 학습 단계와 불일치 발생
- 학습을 위해 데이터가 변경된 경우
    - 출력층의 다른 위치의 벡터들은 예측과 학습에 사용되지 않는다. -> 손실함수가 무시
    - 변경된 데이터 위치의 출력층 벡터만 사용이 된다.
    - 여러 군데가 변경되어도 전부 예측한다.

2. 다음 문장 예측
- 두 개의 문장이 주어지고 두 문장이 이어지는지 아닌지를 맞춘다.
- 5:5 비율로 이어진 문장과 랜덤 문장을 준다.
- [SEP]라는 토큰으로 문장 구분 -> 문장 끝에 붙인다
- [CLS]라는 토큰은 분류 문제를 풀기위한 토큰
- [CLS] 토큰 위치의 출력층에서 이진 분류 문제를 푼다.
- 마스크드 언어 모델, 다음 문장 예측은 같이 학습하여 loss를 합한다.

#### 7) 세그먼트 임베딩
- 문장 구분을 위한 임베딩 층 : 첫번째 문장에선 0, 두번째 문장에선 1
- 총 3개의 임베딩 층 사용
    - WordPiece 임베딩 : 실질적 입력 워드 임베딩, 벡터 종류는 단어집합의 크기 30522개
    - Position 임베딩 : 위치 정보 학습. 벡터 종류는 문장 최대 길이 512개
    - Segment 임베딩 : 두 개의 문장 구분. 벡터 종류는 문장 최대 개수인 2개
- 여기서 말하는 두 개의 문장은 딱 두 문장이 아닐 수 있다.  두개의 단락이 될 수도 있다.
- 두 개의 문장을 받을 필요가 없는 경우 : 0으로만 임베딩을 더해준다.

#### 8) BERT를 파인 튜닝 하기
- 파인 튜닝 : 사전 학습된 BERT에 추가로 학습시키는 것

1. 하나의 텍스트에 대한 텍스트 분류 유형
- 문서 시작에 [CLS] 토큰 입력
- [CLS] 토큰 위치의 출력층에서 밀집층을 추가해 분류 예측

2. 하나의 텍스트에 대한 태깅 작업
- 품사 태깅 작업, 개체명 인식 작업 등
- 입력 텍스트의 각 토큰 위치에 밀집층을 사용해 분류 예측

3. 텍스트의 쌍에 대한 분류 또는 회귀 문제
- 자연어 추론 : 두 문장의 관계 분류. 모순, 함, 중립
- 두 종류의 세그먼트 임베딩을 사용한다.

4. 질의 응답
- 질문, 본문 입력 받을 경우 -> 본문의 일부분을 추출해 답변

#### 9) 어텐션 마스크
- 불필요하게 패딩 토큰에대해 어텐션 하는 것을 방지 : 단어와 패딩 토큰 구분하도록 알려주는 입력
- 1은 실제 단어, 0은 패딩 토큰

<br></br>

#### 1) 마스크드 언어 모델과 토크나이저
- 모델과 토크나이저는 항상 맵핑 관계 : 단어와 정수 정보가 다른 것을 방지
```
from transformers import TFBertForMaskedLM
from transformers import AutoTokenizer
```
- `TFBertForMaskedLM.from_pretrained('BERT 모델 이름')` : [MASK]라고 되어있는 단어를 맞추기 위한 마스크드 언어 모델로 로드
- `AutoTokenizer.from_pretrained('모델 이름')` : 해당 모델 학습 당시 토크나이저 로드

#### 2) BERT의 입력
- 토크나이저로 정수 인코딩 : `tokenizer`
- 정수 인코딩 결과 : `inputs['input_ids']`
- 세그먼트 인코딩 결과 : `inputs['token_type_ids']`
- 어텐션 마스크 : `inputs['attention_mask']`

#### 3) [MASK] 토큰 예측
- `from transformers import FillMaskPipeline` : 모델과 토크나이저 지정하면 예측 결과를 보여줌
- 상위 5개의 단어 출력

<br></br>

#### 1) 마스크드 언어 모델과 토크나이저
- klue/bert-base : 대표적인 한국어 BERT
- BERT 로드 : `from_pt=True` 인자를 주면 파이토치로 학습된 모델이지만 텐서플로우에서 사용하는 것으로 지정

#### 2) BERT의 입력
- 구글 BERT와 동일

#### 3) [MASK] 토큰 예측
- 구글 BERT와 동일

<br></br>

#### 1) 다음 문장 예측 모델과 토크나이저
```
from transformers import TFBertForNextSentencePrediction
from transformers import AutoTokenizer
```
- `TFBertForNextSentencePrediction.from_pretrained('BERT 모델 이름')` : 다음 문장 예측 모델

#### 2) BERT의 입력
- 토크나이저로 두 개의 문장 정수 인코딩 : `tokenizer(첫문장, 다음문장, return_tensors="tf")`
- 마스크드 언어 모델과 동일하게 정수 인코딩 결과는 `input_ids`로 알 수 있다.
- `tokenizer.decode`를 이용해 다시 디코딩을 해 문장을 돌릴 수 있다.

#### 3) 다음 문장 예측하기
- 소프트 맥스 함수를 지나기 전 값 : `model(encoding['input_ids'], token_type_ids=encoding['token_type_ids'])[0]`
- 소프트 맥스 함수 : `tf.keras.layers.Softmax()`
- 확률이 큰 값 리턴 : `tf.math.argmax(probs, axis=-1).numpy()`
- 이어지는 문장은 0, 이어지지 않으면 1

<br></br>

#### 1) 다음 문장 예측 모델과 토크나이저
- 구글 BERT와 동일
- `from_pt` 인자를 `True`로 지정한다.

#### 2) 다음 문장 예측
- 구글 BERT와 동일

<br></br>

#### 1) BERT의 문장 임베딩
- 문장 벡터를 얻는 방법(3가지 소개)
    - [CLS] 토큰의 출력 벡터를 문장 벡터로 간주
    - BERT의 모든 출력 벡터의 평균을 낸다.(평균 풀링) -> 모든 단어의 의미를 반영
    - 맥스 풀링을 진행한 벡터 -> 중요한 단어의 의미를 반영

#### 2) SBERT
- BERT의 문장 임베딩 성능을 개선

1. 문장 쌍 분류 태스크로 파인 튜닝
- NLI 문제를 푸는 것 : 두 문장이 수반, 모순, 중립 등의 어떤 관계인지 예측
- 문장 각각을 BERT의 입력으로 넣어 평균 풀링 또는 맥스 풀링 -> 문장 임베딩 벡터 얻음
- 두 문장 임베딩 벡터의 차이를 구하고 세 가지 벡터를 연결
- 연결한 벡터를 출력층으로 보냄 -> 다중 클래스 분류 문제
   : 분류하고자 하는 클래스 갯수가 K -> 3n by K 크기의 행렬을 곱한 뒤 소프트 맥스 함수 통과

2. 문장 쌍 회귀 태스크로 파인 튜닝
- STS 문제 : 두 개의 문장으로부터 의미적 유사성 구하는 문제
- 유사도 범위 : 0~5
- 각 문장을 BERT의 입력으로 넣어 풀링 -> 문장 임베딩 벡터
- 두 문장 임베딩 벡터의 코사인 유사도 구함 -> 레이블 유사도와의 평균 제곱 오차를 최소화하도록 학습

---

- 참고
>> 딥 러닝을 이용한 자연어 처리 입문, 유원준, 안상준
      * 위 글 내용은 이 책의 내용을 공부하여 정리한 것입니다.
